src/main/scala/org/apache/spark/mllib/linalg/distributed/BetterBlockMatrix.scala

```diff
@@ -0,0 +1,183 @@
+package org.apache.spark.mllib.linalg.distributed
+
+import org.apache.spark._
+import org.apache.spark.rdd._
+import org.apache.spark.mllib.linalg._
+import org.apache.spark.mllib.linalg.distributed._
+
+object BetterBlockMatrix {
+
+  class BlockMatrixMultiplyRDD(l: BlockMatrix, r: BlockMatrix, deps: Seq[Dependency[_]], partitions: Array[Partition])
+      extends RDD[((Int, Int), Matrix)](l.blocks.sparkContext, deps) {
+
+    val lPartitioner = l.blocks.partitioner.get.asInstanceOf[GridPartitioner]
+    val rPartitioner = r.blocks.partitioner.get.asInstanceOf[GridPartitioner]
+    val nRows: Long = l.numRows()
+    val nCols: Long = r.numCols()
+    val rowsPerBlock: Int = l.rowsPerBlock
+    val colsPerBlock: Int = r.colsPerBlock
+    val rowBlocks = l.numRowBlocks
+    val colBlocks = r.numColBlocks
+    val nProducts = l.numColBlocks
+    val rowsRemainder: Int = (nRows % rowsPerBlock).toInt
+    val colsRemainder: Int = (nCols % colsPerBlock).toInt
+
+    private def toDenseMatrix(x: Matrix): DenseMatrix = x match {
+      case x: DenseMatrix => x
+      case x: SparseMatrix => x.toDense
+    }
+
+    private def fmaDenseMatricesWithExtension(x: DenseMatrix, y: DenseMatrix, result: DenseMatrix) {
```

Comment:
For benefit of future maintainers, maybe add a comment here explaining what this does? fma is not an often used abbreviation, and it'd be nice to not have to work exactly what intended purpose of this is.